SFO Public Department, referred to as SFO, has captured the salary data of all its employees for the years 2011-2014. Now in 2018, the organization is facing some financial crisis. As a first step, HR wants to rationalize employee cost to save payroll budget. You have to perform data manipulation and answer the questions below:

1. How much total salary cost has increased from year 2011 to 2014?
2. Who was the top-earning employee across all the years?
- Objective: Perform data manipulation and visualization techniques.

In [16]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np

In [2]:
df = pd.read_csv('Salaries.csv')
df.shape

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (12) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


(148648, 13)

In [3]:
df.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

In [4]:
df['TotalPay'].groupby(df['Year']).sum()

Year
2011    2.594113e+09
2012    2.724736e+09
2013    2.918656e+09
2014    2.876911e+09
Name: TotalPay, dtype: float64

### How much total salary cost has increased from year 2011 to 2014?
2.876911 - 2.594113 = 0.282798+09 ~ 289,000,000
According to the output above total salary cost has increased by about 289 million from 2011-2014.

### Who was the top-earning employee across all the years?

In [13]:
employees = []

In [14]:
for empl in df['EmployeeName'].unique():
    employees.append(empl)
print(employees)

['NATHANIEL FORD', 'GARY JIMENEZ', 'ALBERT PARDINI', 'CHRISTOPHER CHONG', 'PATRICK GARDNER', 'DAVID SULLIVAN', 'ALSON LEE', 'DAVID KUSHNER', 'MICHAEL MORRIS', 'JOANNE HAYES-WHITE', 'ARTHUR KENNEY', 'PATRICIA JACKSON', 'EDWARD HARRINGTON', 'JOHN MARTIN', 'DAVID FRANKLIN', 'RICHARD CORRIEA', 'AMY HART', 'SEBASTIAN WONG', 'MARTY ROSS', 'ELLEN MOFFATT', 'VENUS AZAR', 'JUDY MELINEK', 'GEORGE GARCIA', 'VICTOR WYRSCH', 'JOSEPH DRISCOLL', 'GREGORY SUHR', 'JOHN HANLEY', 'RAYMOND GUZMAN', 'DENISE SCHMITT', 'MONICA FIELDS', 'HARLAN KELLY-JR', 'DAVID SHINN', 'GARY AMELIO', 'JOHN TURSI', 'JOSE VELO', 'KHOA TRINH', 'SUSAN CURRIN', 'JAMES BOSCH', 'JONES WONG', 'BRENDAN WARD', 'MICHAEL THOMPSON', 'THOMAS ABBOTT', 'THOMAS SIRAGUSA', 'SHARON MCCOLE WICHER', 'EDWIN LEE', 'BRYAN RUBENSTEIN', 'TRENT RHORER', 'JAMES DUDLEY', 'KEN YEE', 'KIRK RICHARDSON', 'BARBARA GARCIA', 'MICHAEL ROLOVICH', 'DARRYL HUNTER', 'RAY CRAWFORD', 'KEVIN CASHMAN', 'BENJAMIN ROSENFIELD', 'MARK KEARNEY', 'DOUGLAS RIBA', 'AI-KYUNG CH

In [17]:
unique_employees = np.unique(employees)
print(unique_employees)

['A BERNARD FATOOH' 'A Bernard  Fatooh' 'A Bernard Fatooh' ...
 'Zuna T Luu' 'Zuri  Jones' 'Zuri Jones']


In [21]:
print(dir(employees))

['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'clear', 'copy', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']


In [30]:
print(dir(np))

['ALLOW_THREADS', 'AxisError', 'BUFSIZE', 'CLIP', 'ComplexWarning', 'DataSource', 'ERR_CALL', 'ERR_DEFAULT', 'ERR_IGNORE', 'ERR_LOG', 'ERR_PRINT', 'ERR_RAISE', 'ERR_WARN', 'FLOATING_POINT_SUPPORT', 'FPE_DIVIDEBYZERO', 'FPE_INVALID', 'FPE_OVERFLOW', 'FPE_UNDERFLOW', 'False_', 'Inf', 'Infinity', 'MAXDIMS', 'MAY_SHARE_BOUNDS', 'MAY_SHARE_EXACT', 'MachAr', 'ModuleDeprecationWarning', 'NAN', 'NINF', 'NZERO', 'NaN', 'PINF', 'PZERO', 'RAISE', 'RankWarning', 'SHIFT_DIVIDEBYZERO', 'SHIFT_INVALID', 'SHIFT_OVERFLOW', 'SHIFT_UNDERFLOW', 'ScalarType', 'Tester', 'TooHardError', 'True_', 'UFUNC_BUFSIZE_DEFAULT', 'UFUNC_PYVALS_NAME', 'VisibleDeprecationWarning', 'WRAP', '_NoValue', '_UFUNC_API', '__NUMPY_SETUP__', '__all__', '__builtins__', '__cached__', '__config__', '__dir__', '__doc__', '__file__', '__getattr__', '__git_revision__', '__loader__', '__mkl_version__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_add_newdoc_ufunc', '_distributor_init', '_globals', '_mat', '_pytes

In [77]:
df.loc[df['EmployeeName'] == 'Kevin Lee'].groupby(df['TotalPay']).max()

Id EmployeeName                        JobTitle    BasePay  \
TotalPay                                                                    
68940.44    55120    Kevin Lee              IS Administrator 2   68940.44   
69187.63    53932    Kevin Lee                Transit Operator   57397.32   
77465.73    52234    Kevin Lee  Senior Parking Control Officer   66774.03   
82242.40   124659    Kevin Lee                Transit Operator   67230.30   
83382.00    50124    Kevin Lee               Personnel Analyst   83382.00   
85487.78    49440    Kevin Lee          Deputy Court Clerk III   84512.34   
85982.31    49271    Kevin Lee          Deputy Court Clerk III   84512.32   
88353.01   124166    Kevin Lee               Personnel Analyst   88353.01   
92619.03   123142    Kevin Lee  IT Operations Support Admin IV   91606.03   
122821.42   42511    Kevin Lee  Electrical Transit System Mech   79220.54   
141023.19   40571    Kevin Lee                Police Officer 3  117171.41   
148703.74   39716    Kevin Lee                Police Officer 3  115866.90   
172558.95  112386    Kevin Lee                      Sergeant 3  137982.14   

           OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
TotalPay                                                                        
68940.44          0.00      0.00  30650.48   68940.44          99590.92  2012   
69187.63       9989.83   1800.48  34577.54   69187.63         103765.17  2012   
77465.73       9599.08   1092.62  31902.38   77465.73         109368.11  2012   
82242.40      10961.25   4050.85  36777.83   82242.40         119020.23  2014   
83382.00          0.00      0.00  35210.59   83382.00         118592.59  2012   
85487.78          0.00    975.44  35902.02   85487.78         121389.80  2012   
85982.31          0.00   1469.99  36080.30   85982.31         122062.61  2012   
88353.01          0.00      0.00  33172.33   88353.01         121525.34  2014   
92619.03          0.00   1013.00  34109.61   92619.03         126728.64  2014   
122821.42     40247.39   3353.49  38781.04  122821.42         161602.46  2012   
141023.19     19606.88   4244.90  34610.10  141023.19         175633.29  2012   
148703.74     23523.30   9313.54  34906.20  148703.74         183609.94  2012   
172558.95     18537.19  16039.62  43039.12  172558.95         215598.07  2014   

           Notes         Agency Status  
TotalPay                                
68940.44     NaN  San Francisco    NaN  
69187.63     NaN  San Francisco    NaN  
77465.73     NaN  San Francisco    NaN  
82242.40     NaN  San Francisco     FT  
83382.00     NaN  San Francisco    NaN  
85487.78     NaN  San Francisco    NaN  
85982.31     NaN  San Francisco    NaN  
88353.01     NaN  San Francisco     FT  
92619.03     NaN  San Francisco     FT  
122821.42    NaN  San Francisco    NaN  
141023.19    NaN  San Francisco    NaN  
148703.74    NaN  San Francisco    NaN  
172558.95    NaN  San Francisco     FT

In [51]:
df['TotalPay'].groupby(df['EmployeeName'] == 'Kevin Lee').sum()

EmployeeName
False    1.111310e+10
True     1.318768e+06
Name: TotalPay, dtype: float64